File: milestone 2/python_modules/module01/ex6/ft_garden_analytics.py
class Plant:
    def __init__(self, name, height):
        self.name = name
        self.height = height
        self.growth = 0

    def grow(self):
        self.growth += 1
        print(f"{self.name} grew 1cm")

    def get_height(self):
        return self.height + self.growth

    def get_growth(self):
        return self.growth

    def get_points(self):
        return self.height + self.growth

    def info(self):
        print(f"- {self.name}: {self.get_height()}cm", end="")


class FloweringPlant(Plant):
    def __init__(self, name, height, color):
        super().__init__(name, height)
        self.color = color

    def info(self):
        super().info()
        print(f", {self.color} flowers (blooming)", end="")


class PrizeFlower(FloweringPlant):
    def __init__(self, name, height, color, points):
        super().__init__(name, height, color)
        self.points = points

    def get_points(self):
        return self.get_height() + self.points * 4

    def info(self):
        super().info()
        print(f", prize points: {self.points}")


class Garden:
    def __init__(self, owner):
        self.owner = owner
        self._plants = []

    def add_plant(self, plant):
        print(f"Added {plant.name} to {self.owner}'s garden")
        self._plants.append(plant)

    def grow_plants(self):
        print(f"\n{self.owner} is helping all plants grow...")
        for p in self._plants:
            p.grow()
        print()

    def get_plants(self):
        return self._plants


class GardenManager:
    def __init__(self):
        self._gardens = []

    @staticmethod
    def create_garden_network(gardeners):
        mgr = GardenManager()
        for name in gardeners:
            mgr.add_garden(name)
        return mgr

    def add_garden(self, name):
        self._gardens.append(Garden(name))

    def find_garden(self, name):
        for g in self._gardens:
            if g.owner == name:
                return g
        return None

    def add_plant(self, garden_name, plant):
        g = self.find_garden(garden_name)
        if g:
            g.add_plant(plant)

    def grow_plants(self, garden_name):
        g = self.find_garden(garden_name)
        if g:
            g.grow_plants()

    def get_scores(self):
        parts = []
        for g in self._gardens:
            score = sum(p.get_points() for p in g.get_plants())
            parts.append(f"{g.owner}: {score}")
        print("Garden scores - " + ", ".join(parts))

    def get_gardencount(self):
        print(f"Total gardens managed: {len(self._gardens)}")

    class GardenStats:
        @staticmethod
        def report(garden):
            print(f"\n=== {garden.owner}'s Garden Report ===\n")
            GardenManager.GardenStats.get_plants(garden)
            GardenManager.GardenStats.get_plant_stats(garden)
            GardenManager.GardenStats.get_plant_types(garden)
            GardenManager.GardenStats.validate_heights(garden)
            print()

        @staticmethod
        def get_plants(garden):
            print("Plants in garden:")
            for p in garden.get_plants():
                p.info()
                print()
            print()

        @staticmethod
        def get_plant_stats(garden):
            plants = garden.get_plants()
            total_growth = sum(p.get_growth() for p in plants)
            print(
                f"Plants added: {len(plants)}, "
                f"total growth: {total_growth}cm"
            )

        @staticmethod
        def get_plant_types(garden):
            regular = flowering = prize = 0
            for p in garden.get_plants():
                if isinstance(p, PrizeFlower):
                    prize += 1
                elif isinstance(p, FloweringPlant):
                    flowering += 1
                else:
                    regular += 1
            print(
                f"Plant types: {regular} regular, "
                f"{flowering} flowering, "
                f"{prize} prize flowers"
            )

        @staticmethod
        def validate_heights(garden):
            ok = all(p.get_height() >= 0 for p in garden.get_plants())
            print(f"Height validation test: {ok}")


if __name__ == "__main__":
    print("=== Garden Management System Demo ===\n")

    manager = GardenManager.create_garden_network(["Alice", "Bob"])

    manager.add_plant("Alice", Plant("Oak Tree", 100))
    manager.add_plant("Alice", FloweringPlant("Rose", 25, "red"))
    manager.add_plant("Alice", PrizeFlower("Sunflower", 50, "yellow", 10))
    manager.grow_plants("Alice")

    manager.add_plant("Bob", FloweringPlant("Rose", 92, "red"))

    alice_garden = manager.find_garden("Alice")
    if alice_garden:
        GardenManager.GardenStats.report(alice_garden)

    manager.get_scores()
    manager.get_gardencount()

File: milestone 2/python_modules/module01/ex6/ft_garden_analytics.cpp
#include <iostream>
#include <memory>
#include <string>
#include <vector>

class Plant
{
  public:
	std::string name;
	int height, growth{0};

	Plant(std::string n, int h) : name(std::move(n)), height(h) {}
	virtual ~Plant() = default;
	
	virtual void grow()
	{
		++growth;
		std::cout << name << " grew 1cm\n";
	}

	int get_height() const
	{
		return (height + growth);
	}
	int get_growth() const
	{
		return (growth);
	}
	virtual int get_points() const
	{
		return (height + growth);
	}

	virtual void info() const
	{
		std::cout << "- " << name << ": " << get_height() << "cm";
	}
};

class FloweringPlant : public Plant
{
  public:
	std::string color;

	FloweringPlant(std::string n, int h, std::string c) : Plant(std::move(n),
		h), color(std::move(c))
	{
	}

	void info() const override
	{
		Plant::info();
		std::cout << ", " << color << " flowers (blooming)";
	}
};

class PrizeFlower : public FloweringPlant
{
  public:
	int points;

	PrizeFlower(std::string n, int h, std::string c,
		int p) : FloweringPlant(std::move(n), h, std::string(c)), points(p)
	{
	}

	int get_points() const override
	{
		return get_height() + points * 4;
	}

	void info() const override
	{
		FloweringPlant::info();
		std::cout << ", prize points: " << points << std::endl;
	}
};

class Garden
{
	std::string owner;
	std::vector<std::unique_ptr<Plant>> plants;

	public:
		explicit Garden(std::string o) : owner(std::move(o)) {}

		const std::string& get_owner() const { return owner; }

		void add_plant(std::unique_ptr<Plant> p)
		{
			std::cout << "Added " << p->name << " to " << owner <<"'s garden\n";
			plants.emplace_back(std::move(p));
		}

		void grow_plants()
		{
			std::cout << "\n" << owner << " is helping all plants grow...\n";
			size_t i = 0;
			while (i < plants.size())
			{
				plants[i]->grow();
				++i;
			}
			std::cout << "\n";
		}

		const std::vector<std::unique_ptr<Plant>>& get_plants() const { return plants; }
};

class GardenManager
{
	std::vector<std::unique_ptr<Garden>> gardens;

	public:
	static GardenManager create_garden_network(const std::vector<std::string>& gardeners)
	{
		GardenManager mgr;
		for (const auto& name : gardeners)
		{
			mgr.add_garden(name);
		}
		return mgr;
	}

	void add_garden(const std::string& name)
	{
		gardens.emplace_back(std::make_unique<Garden>(name));
	}

	Garden* find_garden(const std::string& name) const 
	{
		size_t i = 0;
		while (i < gardens.size())
		{
			if (gardens[i]->get_owner() == name) return gardens[i].get();
			++i;
		}
		return nullptr;
	}

	void add_plant(const std::string& garden_name, std::unique_ptr<Plant> p)
	{
		Garden* g = find_garden(garden_name);
		if (g) g->add_plant(std::move(p));
	}

	void grow_plants(const std::string& garden_name)
	{
		Garden* g = find_garden(garden_name);
		if (g) g->grow_plants();
	}

	void get_scores() const
	{
		std::cout << "Garden scores - ";
		size_t i = 0;
		while(i < gardens.size())
		{
			const Garden* g = gardens[i].get();
			int score = 0;
			const auto& vec = g->get_plants();
			size_t j = 0;
			while (j < vec.size())
			{
				score += vec[j]->get_points();
				++j;
			}
			std::cout << g->get_owner() << ": " << score;
			if (i + 1 < gardens.size())
				std::cout << ", ";
			++i;
		}
		std::cout << std::endl;
	}

	void get_gardencount() const
	{
		std::cout << "Total gardens managed: " << gardens.size() << "\n";
	}

	struct GardenStats
	{
		static void report(const Garden& garden)
		{
			std::cout << "\n=== " << garden.get_owner() << "'s Garden Report ===\n\n";
			get_plants(garden);
			get_plant_stats(garden);
			get_plant_types(garden);
			validate_heights(garden);
			std::cout << "\n";
		}

		static void get_plants(const Garden& garden)
		{
			std::cout << "Plants in garden:\n";
			const auto& vec = garden.get_plants();
			size_t i = 0;
			while (i < vec.size())
			{
				vec[i]->info();
				std::cout << "\n";
				++i;
			}
			std::cout << "\n";
		}

		static void get_plant_stats(const Garden& garden)
		{
			int total_growth = 0;
			const auto& vec = garden.get_plants();
			size_t i = 0;
			while (i < vec.size())
			{
				total_growth += vec[i]->get_growth();
				++i;
			}
			std::cout << "Plants added: " << vec.size() << ", total growth: " << total_growth << "cm\n";
		}

		static void get_plant_types(const Garden& garden)
		{
			int regular = 0, flowering = 0, prize = 0;
			const auto& vec = garden.get_plants();
			size_t i = 0;
			while (i < vec.size())
			{
				const Plant* p = vec[i].get();
				if (dynamic_cast<const Plant*>(p) && !dynamic_cast<const FloweringPlant*>(p))
					++regular;
				if (dynamic_cast<const FloweringPlant*>(p) && !dynamic_cast<const PrizeFlower*>(p))
					++flowering;
				if (dynamic_cast<const PrizeFlower*>(p))
					++prize;
				++i;
			}
			std::cout << "Plant types: " << regular << " regular, " << flowering << " flowering, " << prize << " prize flowers\n";
		}

		static void validate_heights(const Garden& garden)
		{
			bool ok = true;
			const auto& vec = garden.get_plants();
			size_t i = 0;
			while (i < vec.size())
			{
				if (vec[i]->get_height() < 0) { ok = false; break; }
				++i;
			}
			std::cout << "Height validation test: ";
			if (ok)
				std::cout << "True";
			else
				std::cout << "False";
			std::cout << std::endl;
		}
	};
};

int	main(void)
{
	std::cout << "=== Garden Management System Demo ===\n\n";
	
	GardenManager manager = GardenManager::create_garden_network({"Alice", "Bob"});

	manager.add_plant("Alice", std::make_unique<Plant>("Oak Tree", 100));
	manager.add_plant("Alice", std::make_unique<FloweringPlant>("Rose", 25, "red"));
	manager.add_plant("Alice", std::make_unique<PrizeFlower>("Sunflower", 50, "yellow", 10));
	manager.grow_plants("Alice");

	manager.add_plant("Bob", std::make_unique<FloweringPlant>("Rose", 92, "red"));

	Garden* aliceGarden = manager.find_garden("Alice");
	if (aliceGarden) GardenManager::GardenStats::report(*aliceGarden);

	manager.get_scores();
	manager.get_gardencount();

	return 0;
}